### PR TITLE
Directly allocate warnings asVectorEnsoObjects

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Directly allocate a `Vector` via `asVectorEnsoObjects` to avoid overhead of a `Vector` delegating to an `Array`.

### Checklist

Please ensure that the following checklist has been satisfied before submitting the PR:

- [x] All code follows the
      [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md),
- [x] Unit tests continue to pass
- [ ] Benchmarks are improved

### CODE REVIEW

File: engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java [hunk 1]
```diff
@@ -38,10 +38,15 @@ public void initializeBenchmark(BenchmarkParams params) throws Exception {
         import Standard.Base.Data.Vector.Builder
         import Standard.Base.Data.Vector.Vector
         import Standard.Base.Data.Array_Proxy.Array_Proxy
+        from Standard.Base.Data.Boolean import False
 
         avg arr =
-            sum acc i = if i == arr.length then acc else
-                @Tail_Call sum (acc + arr.at i) i+1
+            sum acc i =
+                stop = i == arr.length
```
Comment (GregoryTravis) [inline engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java:45]: What effect do these changes have?

**Review by GregoryTravis — APPROVED** [review]

File: engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime-benchmarks/src/main/java/org/enso/interpreter/bench/benchmarks/semantic/VectorBenchmarks.java:45]: It is easier to analyze the IGV graphs. Particularly one can select line 45 and ask for _show all nodes related to this line_. Otherwise I wouldn't be able to generate [these two screenshots](https://github.com/enso-org/enso/pull/10079#issuecomment-2140642492):

1ms case
8ms case

**Comment by JaroslavTulach** [comment]
[engine benchmarks](https://github.com/enso-org/enso/actions/runs/9308413771) are broken [averageOverSlice](https://github.com/enso-org/enso/actions/runs/9308413771/job/25621772054#step:7:5287) yields a `NPE`. Let's [try again](https://github.com/enso-org/enso/actions/runs/9313164586). Another [engine benchmarks run](https://github.com/enso-org/enso/actions/runs/9318760162) is needed. [averageOverSlice failed again](https://github.com/enso-org/enso/actions/runs/9313164586/job/25635135397#step:7:5286).

First [stdlib benchmarks](https://github.com/enso-org/enso/actions/runs/9308421283) failed scheduling [another run](https://github.com/enso-org/enso/actions/runs/9312697673).

**Comment by JaroslavTulach** [comment]
After recent changes I seem to have eliminated all `if` statements in the 
```
sbt:runtime-benchmarks> benchOnly VectorBenchmarks.averageOverSlice
[info] Benchmark                          Mode  Cnt  Score    Error  Units
[info] VectorBenchmarks.averageOverSlice  avgt    5  0.001 ±  0.001  ms/op
```
e.g. the benchmark is now as **fast as it should be**. Let's confirm it by another round of [engine benchmarks](https://github.com/enso-org/enso/actions/runs/9313164586).

**Comment by JaroslavTulach** [comment]
### Note on debugging

After running
```
sbt:runtime-benchmarks> benchOnly VectorBenchmarks.averageOverSlice
```
there is a file writen to disk `engine/runtime-benchmarks/target/bench-data/averageOverSlice.enso` which represents the exact source used by the benchmark. Add a `main` function to its bottom:
```ruby
main =
    v = fibarr 1000 1000000
    avg v
```
and you can run it and `--debug` it as regular Enso file. E.g.:
```
sbt:runtime-benchmarks> enso/runEngineDistribution --run engine/runtime-benchmarks/target/bench-data/averageOverSlice.enso
```
Btw. @radeusgd, is there a way to load this file in and then use its functions from REPL? I've never figured out how to do both - load a file and then REPL...

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java [hunk 2]
```diff
@@ -41,64 +44,87 @@ boolean isEmpty() {
...
-    } else if (primitiveArray instanceof long[] longArray) {
-      if (e instanceof Long l) {
-        if (size == longArray.length) {
+    }
+    addImpl(e);
+  }
+
+  private void addImpl(Object e) {
+    while (true) {
+      // loop at most twice thanks to continue and break
```
Comment (radeusgd) [inline engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java]: I find this continue/break logic hard to follow.

I suggest breaking the logic into smaller functions, allowing much clearer control flow:
```diff
diff --git a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java b/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java
--- a/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java	(revision 4dbb1506eb1344a8ebda2784fd7586cf3d1f1b3e)
+++ b/engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java	(date 1717144249779)
@@ -47,85 +47,98 @@
    * @param warnings library to check for values with warnings
    */
   void add(Object e, WarningsLibrary warnings) {
-    if (!nonTrivialEnsoValue) {
-      if (warnings.hasWarnings(e)) {
-        nonTrivialEnsoValue = true;
-      } else {
-        var isEnsoValue = e instanceof EnsoObject || e instanceof Long || e instanceof Double;
-        if (!isEnsoValue) {
-          nonTrivialEnsoValue = true;
-        }
-      }
-    }
-    addImpl(e);
-  }
+    updateNonTrivialValueCheck(e, warnings);
 
-  private void addImpl(Object e) {
-    while (true) {
-      // loop at most twice thanks to continue and break
-      if (objectArray != null) {
-        if (size == objectArray.length) {
-          CompilerDirectives.transferToInterpreter();
-          objectArray = Arrays.copyOf(objectArray, size * 2);
-        }
-        objectArray[size++] = e;
-      } else if (primitiveArray instanceof long[] longArray) {
-        if (e instanceof Long l) {
-          if (size == longArray.length) {
-            CompilerDirectives.transferToInterpreter();
-            primitiveArray = longArray = Arrays.copyOf(longArray, size * 2);
-          }
-          longArray[size++] = l;
-        } else {
-          CompilerDirectives.transferToInterpreter();
-          objectArray = new Object[longArray.length];
-          for (int i = 0; i < size; i++) {
-            objectArray[i] = longArray[i];
-          }
-          primitiveArray = null;
-          continue;
-        }
-      } else if (primitiveArray instanceof double[] doubleArray) {
-        if (e instanceof Double d) {
-          if (size == doubleArray.length) {
-            CompilerDirectives.transferToInterpreter();
-            primitiveArray = doubleArray = Arrays.copyOf(doubleArray, size * 2);
-          }
-          doubleArray[size++] = d;
-        } else {
-          CompilerDirectives.transferToInterpreter();
-          objectArray = new Object[doubleArray.length];
-          for (int i = 0; i < size; i++) {
-            objectArray[i] = doubleArray[i];
-          }
-          primitiveArray = null;
-          continue;
-        }
-      } else {
-        assert objectArray == null;
-        assert primitiveArray == null;
-        assert size == 0;
-        switch (e) {
-          case Long l -> {
-            var arr = new long[initialCapacity];
-            arr[0] = l;
-            primitiveArray = arr;
-          }
-          case Double d -> {
-            var arr = new double[initialCapacity];
-            arr[0] = d;
-            primitiveArray = arr;
-          }
-          default -> {
-            var arr = new Object[initialCapacity];
-            arr[0] = e;
-            objectArray = arr;
-          }
-        }
-        size = 1;
-      }
-      break;
-    }
+    if (objectArray != null) {
+      addObject(e);
+    } else if (primitiveArray instanceof long[] longArray) {
+      if (e instanceof Long l) {
+        if (size == longArray.length) {
+          CompilerDirectives.transferToInterpreter();
+          primitiveArray = longArray = Arrays.copyOf(longArray, size * 2);
+        }
+        longArray[size++] = l;
+      } else {
+        retypeToObject(longArray);
+        addObject(e);
+      }
+    } else if (primitiveArray instanceof double[] doubleArray) {
+      if (e instanceof Double d) {
+        if (size == doubleArray.length) {
+          CompilerDirectives.transferToInterpreter();
+          primitiveArray = doubleArray = Arrays.copyOf(doubleArray, size * 2);
+        }
+        doubleArray[size++] = d;
+      } else {
+        retypeToObject(doubleArray);
+        addObject(e);
+      }
+    } else {
+      // initialize with first element
+      assert objectArray == null;
+      assert primitiveArray == null;
+      assert size == 0;
+      switch (e) {
+        case Long l -> {
+          var arr = new long[initialCapacity];
+          arr[0] = l;
+          primitiveArray = arr;
+        }
+        case Double d -> {
+          var arr = new double[initialCapacity];
+          arr[0] = d;
+          primitiveArray = arr;
+        }
+        default -> {
+          var arr = new Object[initialCapacity];
+          arr[0] = e;
+          objectArray = arr;
+        }
+      }
+      size = 1;
+    }
+  }
+
+  private void updateNonTrivialValueCheck(Object e, WarningsLibrary warnings) {
+    if (!nonTrivialEnsoValue) {
+      if (warnings.hasWarnings(e)) {
+        nonTrivialEnsoValue = true;
+      } else {
+        var isTrivial = e instanceof EnsoObject || e instanceof Long || e instanceof Double;
+        if (!isTrivial) {
+          nonTrivialEnsoValue = true;
+        }
+      }
+    }
+  }
+
+  private void addObject(Object e) {
+    assert objectArray != null;
+    if (size == objectArray.length) {
+      CompilerDirectives.transferToInterpreter();
+      objectArray = Arrays.copyOf(objectArray, size * 2);
+    }
+    objectArray[size++] = e;
+  }
+
+  private void retypeToObject(long[] longArray) {
+    CompilerDirectives.transferToInterpreter();
+    assert objectArray == null;
+    objectArray = new Object[longArray.length];
+    for (int i = 0; i < size; i++) {
+      objectArray[i] = longArray[i];
+    }
+    primitiveArray = null;
+  }
+
+  private void retypeToObject(double[] doubleArray) {
+    CompilerDirectives.transferToInterpreter();
+    assert objectArray == null;
+    objectArray = new Object[doubleArray.length];
+    for (int i = 0; i < size; i++) {
+      objectArray[i] = doubleArray[i];
+    }
+    primitiveArray = null;
   }
 
   /** Obtains an element from the builder */
@@ -232,9 +245,11 @@
   Object asVector() {
     var res = toArray();
     if (res instanceof long[] longs) {
+      assert !nonTrivialEnsoValue;
       return Vector.fromLongArray(longs);
     }
     if (res instanceof double[] doubles) {
+      assert !nonTrivialEnsoValue;
       return Vector.fromDoubleArray(doubles);
     }
     if (nonTrivialEnsoValue) {
```

I think it is easier to understand if there is no loops.

**Comment by radeusgd** [comment]
> Btw. @radeusgd, is there a way to load this file in and then use its functions from REPL? I've never figured out how to do both - load a file and then REPL...

The `--repl` option is essentially a wrapper for `Debug.breakpoint`, importing `Standard.Base`. So if you want a REPL with additional imports, what I usually do is create a file:
```
from Standard.Base import all
import Standard.Base.Runtime.Debug
from Somewhere import Something

and_additional_helper_defs = ...

main = Debug.breakpoint
```
and run it - this gives me a REPL with additional customization.

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java]: Previously there was a recursive call. But I realized that Truffle PE doesn't like recursive calls (as they cannot be inlined indefinitely). As such I returned back to plain old `goto` (via `while`, `continue` and `break`). It is way more easier for the compiler to process that.

Comment (radeusgd) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java]: But my suggested solution does not have a recursive call. Shouldn't the compiler be able to inline `addObject`?

**Review by Akirathan — APPROVED** [review]

File: engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/interpreter/runtime/data/vector/ArrayBuilder.java]: `addObject` is a nice refactoring. Yes, inlining shall work fine. Applied in 52f815b

**Comment by JaroslavTulach** [comment]
Another [engine benchmarks run](https://github.com/enso-org/enso/actions/runs/9318760162) is needed. [averageOverSlice failed again](https://github.com/enso-org/enso/actions/runs/9313164586/job/25635135397#step:7:5286).

**Comment by JaroslavTulach** [comment]
The final benchmarks results aren't spectacular, but they are also not worse. The most important benchmark in question:

image

seems to historically oscillate between 1ms and ~8ms. Looks like that sometimes it [splits and sometimes it doesn't](https://github.com/enso-org/enso/pull/10079#issuecomment-2141335183). Locally I used to have slower ~8ms results, but now after the fix, I am getting 1ms result consistently.

Let's integrate and hope for stable `averageOverSlice` & co.

**Review by hubertp — COMMENTED** [review]
fwiw lgtm
